### PR TITLE
hpke: Exposing KDF API for Suite.

### PR DESCRIPTION
**Pull request by armfazh**
Exposes abstract and easier to use KDF API.
cc: @chris-wood @cjpatton

### CODE REVIEW

**Review by wbl — APPROVED** [review]
Should we do the same for the KEM and AEAD or does that make less sense?

File: hpke/algs.go [hunk 1]
```diff
@@ -44,7 +45,7 @@ func (k KEM) IsValid() bool {
 	}
 }
 
-func (k KEM) Scheme() kem.AuthScheme {
+func (k KEM) Scheme() kem.Scheme {
```
Comment (cjpatton) [inline hpke/algs.go]: What's the reason for this change?

File: hpke/hpke.go [hunk 2]
```diff
@@ -282,6 +290,7 @@ var (
 	errInvalidKDF             = errors.New("hpke: invalid KDF identifier")
 	errInvalidKEM             = errors.New("hpke: invalid KEM identifier")
 	errInvalidAEAD            = errors.New("hpke: invalid AEAD identifier")
+	errNotAuthenticatedKEM    = errors.New("hpke: not authenticated KEM")
```
Comment (cjpatton) [inline hpke/hpke.go]: This error is a bit cryptic. What does it mean?

File: hpke/algs.go [hunk 3]
```diff
@@ -110,14 +111,28 @@ func (k KDF) IsValid() bool {
 	}
 }
 
-func (k KDF) Hash() crypto.Hash {
+// Size returns the size (in bytes) of the Suite.KDFExtract function.
+func (k KDF) Size() int {
```
Comment (cjpatton) [inline hpke/algs.go]: I think this needs a more specific name, like "ExtractSize".

File: hpke/util.go [hunk 4]
```diff
@@ -85,7 +85,21 @@ func (suite Suite) getSuiteID() (id [10]byte) {
...
+func (suite Suite) KDFExtract(secret, salt []byte) (pseudorandomKey []byte) {
+	return hkdf.Extract(suite.kdfID.hash(), secret, salt)
+}
+
+func (suite Suite) KDFExpand(pseudorandomKey, info []byte, outputLen uint) []byte {
+	output := make([]byte, outputLen)
+	rd := hkdf.Expand(suite.kdfID.hash(), pseudorandomKey, info)
+	_, err := io.ReadFull(rd, output)
+	if err != nil {
+		return nil
```
Comment (cjpatton) [inline hpke/util.go]: `nil` is ambiguous, since the output length would be 0 if `outputLen == 0`. If this error might be caused by user error, then I would suggest propagating it. If not, let's panic instead.

**Review by cjpatton — COMMENTED** [review]
In review of cloudflare#170 we decided not to add this API. I'm curious what changed your mind?

File: hpke/util.go (hunk 4 above)
Comment (chris-wood) [reply hpke/util.go]: Let's panic. This is an internal error we can't recover from.

Comment (chris-wood) [reply hpke/util.go]: That said, we also don't validate any of the inputs here. The output length, in particular, [needs to be less than the `255*ExtractSize()`](https://tools.ietf.org/html/rfc5869). Should we add that?

File: hpke/util.go [hunk 5]
```diff
@@ -72,22 +73,38 @@ func (st state) verifyPSKInputs(psk, pskID []byte) error {
...
-	binary.BigEndian.PutUint16(id[4:6], uint16(suite.KemID))
-	binary.BigEndian.PutUint16(id[6:8], uint16(suite.KdfID))
-	binary.BigEndian.PutUint16(id[8:10], uint16(suite.AeadID))
+	binary.BigEndian.PutUint16(id[4:6], uint16(suite.kemID))
+	binary.BigEndian.PutUint16(id[6:8], uint16(suite.kdfID))
+	binary.BigEndian.PutUint16(id[8:10], uint16(suite.aeadID))
 	return
 }
 
+func (suite Suite) KDFExtract(secret, salt []byte) (pseudorandomKey []byte) {
```
Comment (chris-wood) [inline hpke/util.go]: I'm curious: why is this a function of the `Suite`, and not a function of the inner `KDF`? [We already expose KDF-specific functions](https://github.com/armfazh/circl/pull/8/files#diff-8b8e4aa268993805b8226304cb9bd0b7c54f4f952d32148952d7ff43af9bb513R115) as part of this change.

File: hpke/hpke.go [hunk 6]
```diff
@@ -242,7 +242,11 @@ func (s *Sender) allSetup(rnd io.Reader) ([]byte, Sealer, error) {
 	case modeBase, modePSK:
 		enc, ss, err = k.EncapsulateDeterministically(s.pkR, seed)
 	case modeAuth, modeAuthPSK:
-		enc, ss, err = k.AuthEncapsulateDeterministically(s.pkR, s.skS, seed)
+		authKEM, ok := k.(kem.AuthScheme)
```
Comment (chris-wood) [inline hpke/hpke.go]: What is this additional validation doing?

**Review by chris-wood — CHANGES_REQUESTED** [review]
Thanks, @armfazh! I'm supportive of exposing the KDF APIs, as they're needed for things like ODoH, but I think this needs a little bit of work before we land it.

File: hpke/util.go (hunk 5 above)
Comment (cjpatton) [reply hpke/util.go]: Good catch.

Comment (armfazh) [reply hpke/util.go]: Updated now

File: hpke/hpke.go (hunk 2 above)
Comment (armfazh) [reply hpke/hpke.go]: better now?

File: hpke/algs.go (hunk 1 above)
Comment (armfazh) [reply hpke/algs.go]: Since Auth* modes are optional, this allows to add, in the future, KEMs that do not support authentication.
Take a look in the updated version, now there are two functions `Scheme` and `AuthScheme`

File: hpke/util.go (hunk 4 above)
Comment (armfazh) [reply hpke/util.go]: added validation of input/output

File: hpke/hpke.go (hunk 6 above)
Comment (armfazh) [reply hpke/hpke.go]: It prevented from using a KEM that doesn't support authentication. (currently all kems do support).
updated now

File: hpke/algs.go (hunk 1 above)
Comment (cjpatton) [reply hpke/algs.go]: Hmm, this seems like it could bite us in the ass in the future. I would suggest we do one of two things. Either
1. We require that HPKE KEMs can be used securely for sender authentication (i.e., there is some sort of peer reviewed analysis of the KEMs security in this setting); or
2. We prevent the user from doing sender authentication unless the KEM is somehow marked as securely supporting this feature.

My preference would be (1.), since it (a) it's simpler and (b) it's likely to be case that any KEM used for encryption is also suitable for sender authentication.

File: hpke/algs.go [hunk 7]
```diff
@@ -110,14 +122,54 @@ func (k KDF) IsValid() bool {
...
+	if len(pseudorandomKey) < k.ExtractSize() {
+		panic(fmt.Errorf("pseudo-random key must be %v bytes", k.ExtractSize()))
+	}
+	maxLength := uint(255 * k.ExtractSize())
+	if outputLen > maxLength {
+		panic(fmt.Errorf("output length must be less than %v bytes", maxLength))
+	}
+	output := make([]byte, outputLen)
+	rd := hkdf.Expand(k.hash(), pseudorandomKey, info)
+	_, err := io.ReadFull(rd, output)
```
Comment (cjpatton) [inline hpke/algs.go:156]: nit: we ought to enforce the output length is as expected.

**Comment by cjpatton** [comment]
Overall I'm supportive of this change. I'll do a full review once there's a PR for cloudlfare/circl.

File: hpke/hpke.go [hunk 8]
```diff
@@ -282,6 +291,7 @@ var (
 	errInvalidKDF             = errors.New("hpke: invalid KDF identifier")
 	errInvalidKEM             = errors.New("hpke: invalid KEM identifier")
 	errInvalidAEAD            = errors.New("hpke: invalid AEAD identifier")
+	errNoAuthKEM              = errors.New("hpke: KEM does not support authentication")
```
Comment (cjpatton) [inline hpke/hpke.go]: `"KEM does not support sender authentication"`

**Comment by armfazh** [comment]
Merged into cloudflare/circl upstream
